### PR TITLE
DHCPv6 service Dynamic DNS fix. Issue #10346

### PR DESCRIPTION
**Pull request by vktg**
- [ ] Redmine Issue: https://redmine.pfsense.org/issues/10346
- [ ] Ready for review

1) The DHCPv6 Server & RA page is not allow to enter IPv6 address in DDNS Server IP field (IPv4?)
2) The DHCPv6 Server & RA page DDNS Domain Key name may only contain the characters a-z, A-Z 0-9, '-' and '_'
'=', '@', '.', '^', etc. are special characters and dhcpd doesn't start if config contains it
3) The DynDNS secret must be enclosed in quotes (“secret”), otherwise you will get error message:
```
/services_dhcpv6.php: The command '/usr/local/sbin/dhcpd -6 -user dhcpd -group _dhcp -chroot 
/var/dhcpd -cf /etc/dhcpdv6.conf -pf /var/run/dhcpdv6.pid vtnet2' returned exit code '1', the output 
was 'Internet Systems Consortium DHCP Server 4.4.1 Copyright 2004-2018 Internet Systems 
Consortium. All rights reserved. For info, please visit https://www.isc.org/software/dhcp/ 
/etc/dhcpdv6.conf line 27: partial base64 value left over: 3. \x09secret 123; ^ Configuration file 
errors encountered -- exiting If you think you have received this message due to a bug rather than a 
configuration issue please read the section on submitting bugs on either our web page at 
www.isc.org or in the README file before submitting a bug. These pages explain the proper 
process and the information we find helpful for debugging. exiting.'

```

### CODE REVIEW

File: src/etc/inc/services.inc [hunk 1]
```diff
@@ -1274,7 +1274,7 @@ function dhcpdkey($dhcpifconf) {
 		$algorithm = empty($dhcpifconf['ddnsdomainkeyalgorithm']) ? 'hmac-md5' : $dhcpifconf['ddnsdomainkeyalgorithm'];
 		$dhcpdconf .= "key {$dhcpifconf['ddnsdomainkeyname']} {\n";
 		$dhcpdconf .= "	algorithm {$algorithm};\n";
-		$dhcpdconf .= "	secret {$dhcpifconf['ddnsdomainkey']};\n";
+		$dhcpdconf .= "	secret \"{$dhcpifconf['ddnsdomainkey']}\";\n";
```
Comment (jim-p) [inline src/etc/inc/services.inc]: The value inside should have double quotes escaped, or else you end up with another situation like https://redmine.pfsense.org/issues/10295

File: src/usr/local/www/services_dhcpv6.php [hunk 2]
```diff
@@ -313,8 +313,11 @@ function dhcpv6_apply_changes($dhcpdv6_enable_changed) {
 	if ($_POST['ddnsupdate'] && !is_domain($_POST['ddnsdomain'])) {
 		$input_errors[] = gettext("A valid domain name must be specified for the dynamic DNS registration.");
 	}
-	if ($_POST['ddnsupdate'] && !is_ipaddrv4($_POST['ddnsdomainprimary'])) {
-		$input_errors[] = gettext("A valid primary domain name server IPv4 address must be specified for the dynamic domain name.");
+	if ($_POST['ddnsupdate'] && !is_ipaddrv6($_POST['ddnsdomainprimary'])) {
```
Comment (jim-p) [inline src/usr/local/www/services_dhcpv6.php]: Does this have to be IPv6? It's a DNS server address, it shouldn't have to match the type of address the daemon is handling. It should probably accept either type in either page (IPv4 or IPv6 on DHCPv4 and DHCPv6). May need testing to confirm that it works in both of those cases.

File: src/usr/local/www/services_dhcpv6.php [hunk 3]
```diff
@@ -800,8 +803,8 @@ function dhcpv6_apply_changes($dhcpdv6_enable_changed) {
 	'ddnsdomainprimary',
 	'DDNS Server IP',
 	$pconfig['ddnsdomainprimary'],
-	'V4'
-))->setHelp('Enter the primary domain name server IPv4 address for the dynamic domain name.');
+	'V6'
+))->setHelp('Enter the primary domain name server IPv6 address for the dynamic domain name.');
```
Comment (jim-p) [inline src/usr/local/www/services_dhcpv6.php]: If the validation test gets changed as suggested, change this to match.

File: src/etc/inc/services.inc (hunk 1 above)
Comment (vktg) [reply src/etc/inc/services.inc]: This is incorrect too.
dnssec-keygen should be used, 
dhcpd.conf(5):
>       You should choose your own secret key, of course.  The ISC BIND 9  dis-
>        tribution  comes	 with  a  program  for	generating  secret keys	called
>        dnssec-keygen.  If you are using	BIND 9's dnssec-keygen,	the above  key
>        would be	created	as follows:
> 
> 	    dnssec-keygen -a HMAC-MD5 -b 128 -n	USER DHCP_UPDATER
> 
>        The  key	 name, algorithm, and secret must match	that being used	by the
>        DNS server.

I tried to use hash_hmac(), but it needs testing

File: src/usr/local/www/services_dhcpv6.php (hunk 2 above)
Comment (vktg) [reply src/usr/local/www/services_dhcpv6.php]: True,
it uses primary6 and secondary6 for IPv6 addresses,
see dhcpd.conf(5):
```
       The primary statement specifies the IP address of the name server whose
       zone  information  is to	be updated.  In	addition to the	primary	state-
       ment there are also the primary6	, secondary and	secondary6 statements.
       The  primary6  statement	specifies an IPv6 address for the name server.
```

File: src/usr/local/www/services_dhcpv6.php (hunk 3 above)
Comment (vktg) [reply src/usr/local/www/services_dhcpv6.php]: changed to "BOTH"

File: src/etc/inc/services.inc (hunk 1 above)
Comment (jim-p) [reply src/etc/inc/services.inc]: That would be to generate a new key, but in this case the user would be entering a key generated elsewhere (on the actual name server) so it has to be a text input.

Comment (vktg) [reply src/etc/inc/services.inc]: all done
not it fixes DynDNS key using strict base64_decode()

**Comment by vktg** [comment]
> I wonder if it is OK to mix IPv4 and IPv6 (e.g. `primary` with `secondary6`). If not we may have to protect against that. Or maybe if there is one IPv4 and one IPv6 should it be `primary` and `primary6` instead? Needs some testing to determine.

I was checked code, and found that there is only one case when more than one DNS server is used:
https://github.com/pfsense/pfsense/blob/2c9e1e43f3619af39dc12c67830cc63a08ac8829/src/etc/inc/services.inc#L780

another question,-
for some reason ddnsdomainprimary is interpreted as an array::
`$newzone['dns-servers'] = array($dhcpifconf['ddnsdomainprimary']);`
https://github.com/pfsense/pfsense/blob/2c9e1e43f3619af39dc12c67830cc63a08ac8829/src/etc/inc/services.inc#L1281
but WebGUI doesn't allow you to enter more than one DNS server.
Should it?

**Comment by jim-p** [comment]
> another question,-
> for some reason ddnsdomainprimary is interpreted as an array::
> 
> https://github.com/pfsense/pfsense/blob/2c9e1e43f3619af39dc12c67830cc63a08ac8829/src/etc/inc/services.inc#L1281

I'm not sure why that is, but it is similar to `dnsserver` which is marked as a list tag which would be treated as an array by the XML parser. Perhaps in the past it overlapped that but eventually changed and that workaround didn't get removed.

> but WebGUI doesn't allow you to enter more than one DNS server.
> Should it?

If the DHCP daemon supports using more than one server we should accommodate it, if it's feasible.

**Comment by vktg** [comment]
Now DHCP/DHCPv6 can use two DDNS servers,
A DDNS pair can have a different IP address family - the dhcpd daemon starts normally

but I found that most of the DDNS options for static entries are ignored:
```
if (isset($sm['ddnsupdate'])) {
					if (($sm['ddnsdomain'] <> "") && ($sm['ddnsdomain'] != $dhcpifconf['ddnsdomain'])) {
						$smdnscfg .= "		ddns-domainname \"{$sm['ddnsdomain']}\";\n";
					}
					$smdnscfg .= "		ddns-update-style interim;\n";
				}
```
https://github.com/pfsense/pfsense/blob/b943d20dcd9a580c18ce804b47f512855272f1dd/src/etc/inc/services.inc#L1183

File: src/usr/local/www/services_dhcp.php [hunk 4]
```diff
@@ -386,14 +387,25 @@ function validate_partial_mac_list($maclist) {
...
+			$input_errors[] = gettext("A valid secondary domain name server IP address must be specified for the dynamic domain name.");
+		}
+		if (!$_POST['ddnsdomainkeyname'] || !$_POST['ddnsdomainkeyalgorithm'] || !$_POST['ddnsdomainkey']) {
+			$input_errors[] = gettext("A valid domain key name, algorithm and secret must be specified.");
+		}
+		if (preg_match('/[^A-Za-z0-9\-_]/', $_POST['ddnsdomainkeyname'])) {
+			$input_errors[] = gettext("The domain key name may only contain the characters a-z, A-Z, 0-9, '-' and '_'");
+		}
+		if ($_POST['ddnsdomainkey'] && !base64_decode($_POST['ddnsdomainkey'], true)) {
+			$input_errors[] = gettext("The domain key secret must be Base64 encoded value.");
```
Comment (jim-p) [inline src/usr/local/www/services_dhcp.php:407]: This should either be
* "The domain key secret must be **a** Base64 encoded value."
or
* "The domain key secret must be Base64 encoded."

File: src/usr/local/www/services_dhcp_edit.php [hunk 5]
```diff
@@ -274,12 +274,22 @@ function staticmaps_sort($ifgui) {
...
+			$input_errors[] = gettext("A valid domain name must be specified for the dynamic DNS registration.");
+		}
+		if (!is_ipaddr($_POST['ddnsdomainprimary'])) {
+			$input_errors[] = gettext("A valid primary domain name server IP address must be specified for the dynamic domain name.");
+		}
+		if (preg_match('/[^A-Za-z0-9\-_]/', $_POST['ddnsdomainkeyname'])) {
+			$input_errors[] = gettext("The domain key name may only contain the characters a-z, A-Z, 0-9, '-' and '_'");
+		}
+		if ($_POST['ddnsdomainkey'] && !base64_decode($_POST['ddnsdomainkey'], true)) {
+			$input_errors[] = gettext("The domain key secret must be Base64 encoded value.");
```
Comment (jim-p) [inline src/usr/local/www/services_dhcp_edit.php:291]: See previous note, this should match.

File: src/usr/local/www/services_dhcpv6.php [hunk 6]
```diff
@@ -310,14 +311,25 @@ function dhcpv6_apply_changes($dhcpdv6_enable_changed) {
...
+			$input_errors[] = gettext("A valid secondary domain name server IP address must be specified for the dynamic domain name.");
+		}
+		if (!$_POST['ddnsdomainkeyname'] || !$_POST['ddnsdomainkeyalgorithm'] || !$_POST['ddnsdomainkey']) {
+			$input_errors[] = gettext("A valid domain key name, algorithm and secret must be specified.");
+		}
+		if (preg_match('/[^A-Za-z0-9\-_]/', $_POST['ddnsdomainkeyname'])) {
+			$input_errors[] = gettext("The domain key name may only contain the characters a-z, A-Z, 0-9, '-' and '_'");
+		}
+		if ($_POST['ddnsdomainkey'] && !base64_decode($_POST['ddnsdomainkey'], true)) {
+			$input_errors[] = gettext("The domain key secret must be Base64 encoded value.");
```
Comment (jim-p) [inline src/usr/local/www/services_dhcpv6.php:331]: See previous note, this should match.

File: src/usr/local/www/services_ntpd_acts.php [hunk 7]
```diff
@@ -0,0 +1,317 @@
+<?php
```
Comment (jim-p) [inline src/usr/local/www/services_ntpd_acts.php:1]: This file doesn't seem to belong to this PR

Comment (vktg) [reply src/usr/local/www/services_ntpd_acts.php]: all fixed

File: src/usr/local/www/services_dhcp_edit.php (hunk 5 above)
Comment (jim-p) [reply src/usr/local/www/services_dhcp_edit.php:291]: This one didn't get fixed.

**Comment by vktg** [comment]
all fixed now

most DynDNS options for static mappings are ignored by `services_dhcpdv4_configre()`,
it seems this is just copied block code from services_dhcp.php to services_dhcp_edit.php
I'll create another issue for this
